jobs/build-charms/builder_local.py

```diff
@@ -298,37 +339,53 @@ def release(self, entity: str, artifact: "Artifact", to_channels: List[str]):
         args = [entity, rev_args] + channel_args + resource_rev_args
         self.charmcraft.release(*args)
 
-    def promote(self, charm_entity, from_channel, to_channels):
+    def promote(self, charm_entity, from_channel, to_channels, dry_run):
         self._echo(
             f"Promoting :: {charm_entity:^35} :: from:{from_channel} to: {to_channels}"
         )
         charm_status = [
-            row
-            for row in self.status(charm_entity)
-            if row["Revision"] and f"{row['Track']}/{row['Channel']}" == from_channel
+            # Get all the releases and associted bases
+            (release, mappings.base)
+            # where the track matches the from_channel
+            for tracks in self.status(charm_entity)
+            if from_channel.startswith(tracks.track)
+            # when that track has a base mapping
+            for mappings in tracks.mappings
+            if mappings.base
+            # if the from_channel exact matches, and it's not a tracking release
+            # The charm was really released to this channel
+            for release in mappings.releases
+            if release.channel == from_channel and release.status != "tracking"
```

Comment:
```suggestion
            for track in self.status(charm_entity)
            if from_channel.startswith(track.track)
            # when that track has a base mapping
            for mapping in track.mappings
            if mapping.base
            # if the from_channel exact matches, and it's not a tracking release
            # The charm was really released to this channel
            for release in mapping.releases
            if release.channel == from_channel and release.status != "tracking"
```
Maybe a nitpick, but I feel this looks more intuitive to me. I found myself getting confused about the plural nouns 🤷🏻